### PR TITLE
[SPARK-40593][BUILD][CONNECT] Support user configurable `protoc` and `protoc-gen-grpc-java` executables when building Spark Connect.

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This PR adds a new profile named `-Puser-defined-protoc` to support that users can build and test `connect` module by specifying custom `protoc` and `protoc-gen-grpc-java` executables.

### Why are the changes needed?
As described in SPARK-40593, the latest versions of `protoc` and `protoc-gen-grpc-java` have the minimum version requirements for basic libraries such as `glibc` and `glibcxx`. Because of that it is not possible to compile the `connect `module out of the box on CentOS 6 or CentOS 7. Instead the following error messages is shown:

```
[ERROR] PROTOC FAILED: /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
/home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.18' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
/home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.14' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
/home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.5' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe) 
```

### Does this PR introduce _any_ user-facing change?
No, the way to using official pre-release  `protoc` and `protoc-gen-grpc-java` binary files is activated by default.

### How was this patch tested?

- Pass GitHub Actions
- Manual test on CentOS6u3 and CentOS7u4

```bash
export CONNECT_PROTOC_EXEC_PATH=/path-to-protoc-exe
export CONNECT_PLUGIN_EXEC_PATH=/path-to-protoc-gen-grpc-java-exe
./build/mvn clean install -pl connector/connect -Puser-defined-protoc -am -DskipTests
./build/mvn clean test -pl connector/connect -Puser-defined-protoc 
```

and

```bash
export CONNECT_PROTOC_EXEC_PATH=/path-to-protoc-exe
export CONNECT_PLUGIN_EXEC_PATH=/path-to-protoc-gen-grpc-java-exe
./build/sbt clean "connect/compile" -Puser-defined-protoc
./build/sbt  "connect/test" -Puser-defined-protoc
```

### CODE REVIEW

**Comment by LuciferYang** [comment]
@HyukjinKwon 

As I mentioned in SPARK-40593

The default glibc version in CentOs6 and CentOs7 is 2.12, but glibc 2.14 is required at least when build `connect` module, so the following compilation errors will occur when build `connect`  module on CentOs6&CentOs7:

```
[ERROR] PROTOC FAILED: /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
/home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.18' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
/home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.14' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
/home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.5' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe) 
```

In the current draft, I added a new profile to let the user specify user-defined `protocExecutable` and `pluginExecutable` for workaourd, for exmaple 

```
export CONNECT_PROTOC_EXEC_PATH = pathToProtocExecutable
export CONNECT_PLUGIN_EXEC_PATH = pathToPluginExecutable
mvn clean install -DskipTests connector/connect -am 
```


Currently, only Maven related parts are implemented in draft. Do we accept this way to workaround?

**Comment by HyukjinKwon** [comment]
How do we get the user-defined protobuf executables for `CONNECT_PROTOC_EXEC_PATH` and `CONNECT_PLUGIN_EXEC_PATH` in CentOS 6 and 7? If this is the only way, I am fine but we should probably document it how to build/use.

**Comment by LuciferYang** [comment]
User need to manually compile `protoc-xxx-linux-x86_64.exe` and `protoc-gen-grpc-java-1.47.0-linux-x86_64.exe` can executable on `CentOs6&CentOs7`.

Or pre install the library to a custom directory on the machine, and change the `rpath` and `interpreter` of the official
`protoc-xxx-linux-x86_64.exe` and `protoc-gen-grpc-java-1.47.0-linux-x86_64.exe` to the library path

I choose the latter way and have build pass, this may be complex, but it is easier to perform than upgrading the OS base library.

Yes, documentation is required to tell users how to use this way. Do we also need to tell users how to build `protoc-xxx-linux-x86_64.exe` and `protoc-gen-grpc-java-1.47.0-linux-x86_64.exe` in the document?

**Comment by LuciferYang** [comment]
Let me finish the sbt part first

**Comment by HyukjinKwon** [comment]
cc @grundprinzip @amaliujia FYI

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -109,6 +109,16 @@ object SparkBuild extends PomBuild {
     if (profiles.contains("jdwp-test-debug")) {
       sys.props.put("test.jdwp.enabled", "true")
     }
+    if (profiles.contains("user-defined-pb")) {
```
Comment (LuciferYang) [inline project/SparkBuild.scala:112]: also cc @JiangHaonan

Comment (amaliujia) [reply project/SparkBuild.scala:112]: Is this from the pom file?

**Comment by amaliujia** [comment]
Looks easy to follow!

File: connector/connect/README.md [hunk 2]
```diff
@@ -24,7 +24,30 @@ or
 ```bash
 ./build/sbt -Phive clean package
 ```
-   
+
+### Build with user-defined `protoc` and `protoc-gen-grpc-java`
```
Comment (HyukjinKwon) [inline connector/connect/README.md:28]: I would add an example in which we need this custom compilation. For example, CentOS 5 and 6.

**Review by HyukjinKwon — APPROVED** [review]
Looks pretty good. cc @grundprinzip too FYI

**Comment by amaliujia** [comment]
LGTM assuming this is tested manually (seems to be hard have a UT)

File: project/SparkBuild.scala (hunk 1 above)
Comment (LuciferYang) [reply project/SparkBuild.scala:112]: It is customary that the profile name should be consistent with that in `pom.xml`.

The `profiles` from env `SBT_MAVEN_PROFILES`,

https://github.com/apache/spark/blob/84fabc27d688601feabb42abfc7356cd743b3c38/build/sbt-launch-lib.bash#L98-L102

File: connector/connect/README.md (hunk 2 above)
Comment (LuciferYang) [reply connector/connect/README.md:28]: done

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -109,6 +109,14 @@ object SparkBuild extends PomBuild {
...
```
Comment (grundprinzip) [inline project/SparkBuild.scala:112]:
```suggestion
    if (profiles.contains("user-defined-protoc")) {
```

**Review by grundprinzip — COMMENTED** [review]
Thanks for the contribution. My only suggestion would be to rename the profile to `user-defined-protoc`.

File: connector/connect/README.md [hunk 4]
```diff
@@ -24,7 +24,31 @@ or
...
+
+When the user cannot use the official `protoc` and `protoc-gen-grpc-java` binary files to build the `connect` module in the compilation environment,
+for example, compiling `connect` module on CentOS 6 or CentOS 7 which the default `glibc` version is less than 2.14, we can try to compile and test by 
+specifying the user-defined `protoc` and `protoc-gen-grpc-java` binary files as follows:
+
+```bash
+export CONNECT_PROTOC_EXEC_PATH=/path-to-protoc-exe
+export CONNECT_PLUGIN_EXEC_PATH=/path-to-protoc-gen-grpc-java-exe
+./build/mvn -Phive -Puser-defined-pb clean package
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/README.md]: done

File: connector/connect/README.md [hunk 5]
```diff
@@ -24,7 +24,31 @@ or
...
+```
+
+or
+
+```bash
+export CONNECT_PROTOC_EXEC_PATH=/path-to-protoc-exe
+export CONNECT_PLUGIN_EXEC_PATH=/path-to-protoc-gen-grpc-java-exe
+./build/sbt -Puser-defined-pb clean package
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/README.md]: done

File: connector/connect/pom.xml [hunk 6]
```diff
@@ -371,4 +350,68 @@
       </plugin>
     </plugins>
   </build>
+  <profiles>
+    <profile>
+      <id>official-pb</id>
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/pom.xml]: done

File: project/SparkBuild.scala (hunk 3 above)
Comment (LuciferYang) [reply project/SparkBuild.scala]: done

**Comment by LuciferYang** [comment]
Any other changes? @HyukjinKwon @grundprinzip @amaliujia Thanks ~

**Comment by amaliujia** [comment]
Checking with @grundprinzip to see if there are more comments?

**Comment by grundprinzip** [comment]
Since I cannot directly edit the PR description or title, I would kindly ask you to do the following changes:

Title: [SPARK-40593][BUILD][CONNECT] Support user configurable `protoc` and `protoc-gen-grpc-java` executables when building Spark Connect.

> ### What changes were proposed in this pull request?
> This PR adds a new profile named `-Puser-defined-protoc` to support that users can build and test `connect` module by specifying custom `protoc ` and `protoc-gen-grpc-java` executables.
> 
> ### Why are the changes needed?
> As described in [SPARK-40593](https://issues.apache.org/jira/browse/SPARK-40593), the latest versions of `protoc` and `protoc-gen-grpc-java` have minimum version requirements for basic libraries such as `glibc` and `glibcxx`. Because of that it is not possible to compile the `connect` module out of the box on CentOS 6 or CentOS 7. Instead the following error messages is shown:
> 
> ```
> [ERROR] PROTOC FAILED: /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
> /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.18' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
> /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.14' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe)
> /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.5' not found (required by /home/disk0/spark-source/connect/target/protoc-plugins/protoc-3.21.1-linux-x86_64.exe) 
> ```
> 
> ### Does this PR introduce _any_ user-facing change?
> No, the way to using official pre-release `protoc` and `protoc-gen-grpc-java` binary files is activated by default.
> 
> ### How was this patch tested?
> * Pass GitHub Actions
> * Manual test on CentOS6u3 and CentOS7u4
> 
> ```shell
> export CONNECT_PROTOC_EXEC_PATH=/path-to-protoc-exe
> export CONNECT_PLUGIN_EXEC_PATH=/path-to-protoc-gen-grpc-java-exe
> ./build/mvn clean install -pl connector/connect -Puser-defined-protoc -am -DskipTests
> ./build/mvn clean test -pl connector/connect -Puser-defined-protoc 
> ```
> 
> and
> 
> ```shell
> export CONNECT_PROTOC_EXEC_PATH=/path-to-protoc-exe
> export CONNECT_PLUGIN_EXEC_PATH=/path-to-protoc-gen-grpc-java-exe
> ./build/sbt clean "connect/compile" -Puser-defined-protoc
> ./build/sbt  "connect/test" -Puser-defined-protoc
> ```

**Comment by LuciferYang** [comment]
Both pr title and description have been updated @grundprinzip Thanks ~

**Review by grundprinzip — APPROVED** [review]

**Comment by grundprinzip** [comment]
Thank you!

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by LuciferYang** [comment]
Thansks @HyukjinKwon @grundprinzip @amaliujia  ~

**Comment by jinhai-cloud** [comment]
@LuciferYang ，When I was compiling master, I encountered the following problem. The model is Mac M2. Could you tell me how to mutate successfully?

```
[ERROR] /Users/cs/GitHub/spark/connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [0:0]: /Users/cs/GitHub/spark/connector/connect/common/target/protoc-plugins/protoc-gen-grpc-java-1.56.0-osx-aarch_64.exe: program not found or is not executable
Please specify a program using absolute path or make sure the program is available in your PATH system variable
--grpc-java_out: protoc-gen-grpc-java: Plugin failed with status code 1.
```

**Comment by LuciferYang** [comment]
What's your compile command? I think it's not related to this PR. I'm using M2 Max, and I can build successfully with the default profile.I think M2 doesn't need to use `user-defined-protoc` and `user-defined-protoc` is for older versions of Linux, such as CentOS 6.

**Pull request closed without merging**